stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -526,6 +526,28 @@ LogicalResult refineReturnShape(PatternRewriter& rewriter, OpType op,
   return refineReturnShape(rewriter, op, shape);
 }
 
+struct RefineAllGatherOpPattern : public OpRewritePattern<AllGatherOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AllGatherOp op,
+                                PatternRewriter& rewriter) const override {
+    auto operandType = op.getOperand().getType().cast<ShapedType>();
+    if (!operandType.hasRank())
+      return rewriter.notifyMatchFailure(op, "expected ranked operand type");
+
+    if (op.getChannelHandle() && !op.getUseGlobalDeviceIds())
+      return rewriter.notifyMatchFailure(op, "unsupported strategy");
+    DenseIntElementsAttr replicaGroups = op.getReplicaGroups();
+    if (!replicaGroups || replicaGroups.getType().getRank() != 2)
+      return rewriter.notifyMatchFailure(op, "unsupported replica_groups");
```

Comment:
Not many of the eval functions have comments.. but I don't understand these constraints fully, so comments may be useful for maintenance.

I.e. Why is channel handle with global IDs supported?

Reply:
https://github.com/GleasonK/stablehlo/blob/main/docs/spec.md#all_gather

Found it in spec. Will review shortly.